### PR TITLE
nixos/lxc: apply recommendations from distrobuilder

### PR DESCRIPTION
**Pull request by aanderse**
###### Description of changes

Our `lxc-container` configuration doesn't properly apply `systemd` configuration. Our module configuration incorrectly uses `systemd.extraConfig`, which is **not** applied to **any** `systemd` units. Take a quick peak at some [Arch documentation](https://wiki.archlinux.org/title/systemd#Notifying_about_failed_services) to see an example of how to properly apply configuration across **all** `systemd` units.

Tested with `nixos-generators` and the following NixOS configuration:
```
# assume this file is called nixos.nix then run the following command to import into lxd:
# lxc image import --alias systemd-test (nixos-generate -f lxc-metadata) (nixos-generate -c nixos.nix -f lxc)
{ modulesPath, config, pkgs, lib, ... }:
{
  imports = [
    "${modulesPath}/virtualisation/lxc-container.nix"
  ];

  systemd.packages = [   
    (pkgs.writeTextFile {
      name = "systemd-lxc-service-overrides";
      destination = "/etc/systemd/system/service.d/lxc.conf";
      text = ''
        [Service]
        ProcSubset=all
        ProtectProc=default
        ProtectControlGroups=no
        ProtectKernelTunables=no
        NoNewPrivileges=no
        LoadCredential=

        # Additional settings for privileged containers
        ProtectHome=no
        ProtectSystem=no
        PrivateDevices=no
        PrivateTmp=no
        ProtectKernelLogs=no
        ProtectKernelModules=no
        ReadWritePaths=
      '';
    })
  ];
}
```

Once this is running in `lxc` I used `systemctl cat ...` to inspect that units were inheriting the proper `Service` configuration, as opposed to without this configuration when they were not.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

I'm requesting everyone who is listed as a reviewer of https://github.com/NixOS/nixpkgs/pull/164129 to review this. I suggest we merge this PR first and then rebase #164129. We might even consider this PR for backport... :man_shrugging:

### CODE REVIEW

File: nixos/modules/virtualisation/lxc-container.nix [hunk 1]
```diff
@@ -146,12 +146,31 @@ in
...
+          LoadCredential=
+
+          # Additional settings for privileged containers
+          ProtectHome=no
+          ProtectSystem=no
+          PrivateDevices=no
+          PrivateTmp=no
+          ProtectKernelLogs=no
+          ProtectKernelModules=no
+          ReadWritePaths=
```
[the remark this answers is not in the document]
Comment (aanderse) [reply nixos/modules/virtualisation/lxc-container.nix:170]: Sounds good. I can set the `default` to `false` and then guard behind this option. Something like `virtualisation.lxc.applySettingsForPrivilegedContainer`... Let me know if you have a better name.

:thinking:

[the remark this answers is not in the document]
Comment (adamcstephens) [reply nixos/modules/virtualisation/lxc-container.nix:170]: What about just `virtualisation.lxc.privilegedContainer` ? Either way works for me.

[the remark this answers is not in the document]
Comment (aanderse) [reply nixos/modules/virtualisation/lxc-container.nix:180]: Sounds great :+1:

**Review by mkg20001 — APPROVED** [review]

**Comment by aanderse** [comment]
@ofborg test lxd

**Pull request merged**